application/forms/ContactGroupForm.php

```diff
@@ -0,0 +1,192 @@
+            ->on(TermInput::ON_ADD, $callValidation)
+            ->on(TermInput::ON_SAVE, $callValidation)
+            ->on(TermInput::ON_PASTE, $callValidation);
+        $groupField
+            ->addElement(
+                new TextElement(
+                    'group_name',
+                    [
+                        'label'    => $this->translate('Name'),
+                        'required' => true
+                    ]
+                )
+            )
+            ->addElement($this->termInput);
+
+        // add form actions
+        $buttonSubmit = $this->addElement(
+            new SubmitElement(
+                'submit',
+                [
+                    'label' => $this->translate('Submit')
+                ]
+            )
+        );
+        $buttonCancel = new SubmitElement(
+            'cancel',
+            [
+                'label'          => $this->translate('Cancel'),
+                'class'          => 'btn-cancel',
+                'formnovalidate' => true
+            ]
+        );
+
+        // bind cancel button and add it in front of the submit button
+        $this->registerElement($buttonCancel);
+        $buttonSubmit->getElement('submit')->prependWrapper((new HtmlDocument())->setHtmlContent($buttonCancel));
+    }
+
+    public function hasBeenCancelled(): bool
+    {
+        $btn = $this->getPressedSubmitElement();
+
+        return $btn !== null && $btn->getName() === 'cancel';
+    }
+
+    public function getPartUpdates(): array
+    {
+        $this->ensureAssembled();
+
+        return $this->termInput->prepareMultipartUpdate($this->request);
+    }
+
+    protected function validateTerms($terms): void
+    {
+        $contactTerms = [];
+        foreach ($terms as $term) {
+            /** @var Term $term */
+            if (strpos($term->getSearchValue(), ':') === false) {
+                $term->setMessage($this->translate('Is not a contact'));
+                continue;
+            }
+
+            list($type, $id) = explode(':', $term->getSearchValue(), 2);
+            if ($type === 'contact') {
+                $contactTerms[$id] = $term;
+            }
+        }
+
+        if (! empty($contactTerms)) {
+            $contacts = (Contact::on(Database::get()))
+                ->filter(Filter::equal('id', array_keys($contactTerms)));
+            foreach ($contacts as $contact) {
+                $contactTerms[$contact->id]
+                    ->setLabel($contact->full_name)
+                    ->setClass('contact');
+            }
+        }
+    }
+
+    /**
+     * @return ?int
+     */
+    public function addGroup()
+    {
+        $data = $this->getValues();
+        $groupIdentifier = null;
+
+        if (! array_key_exists('group', $data)) {
+            return null;
+        } elseif (
+            ! array_key_exists('group_name', $data['group'])
+            || ! array_key_exists('group_members', $data['group'])
+        ) {
+            return null;
+        }
+
```

Comment:
You can simplify it:
```suggestion
if (! isset($data['group']['group_name'], $data['group']['group_members'])) {
            return null;
}
```